lib/externalGraderResults.js

```diff
@@ -11,129 +10,116 @@ const externalGradingSocket = require('./externalGradingSocket');
-let QUEUE_URL = null;
+const abortController = new AbortController();
+const processingFinished = deferredPromise();
 
-module.exports.init = function (callback) {
+module.exports.init = async function () {
   // If we're not configured to use AWS, don't try to do anything here
-  if (!config.externalGradingUseAws) {
-    return callback(null);
-  }
+  if (!config.externalGradingUseAws) return;
 
   const sqs = new AWS.SQS(config.awsServiceGlobalOptions);
-  loadQueueUrl(sqs, (err) => {
-    if (ERR(err, callback)) return;
-    callback(null);
-    async.forever((next) => {
-      async.waterfall(
-        [
-          (callback) => {
-            async.doUntil(
-              (done) => {
-                const params = {
-                  MaxNumberOfMessages: 10,
-                  QueueUrl: QUEUE_URL,
-                  WaitTimeSeconds: 20,
-                };
-                sqs.receiveMessage(params, (err, data) => {
-                  if (ERR(err, done)) return;
-                  if (!data.Messages) {
-                    return done(null, null);
-                  }
-                  done(null, data.Messages);
-                });
-              },
-              (messages, callback) => {
-                callback(null, !!messages);
-              },
-              (err, messages) => {
-                if (ERR(err, callback)) return;
-                callback(null, messages);
-              }
-            );
-          },
-          (messages, callback) => {
-            async.each(
-              messages,
-              (message, callback) => {
-                let parsedMessage;
-                let receiptHandle;
-                try {
-                  parsedMessage = JSON.parse(message.Body);
-                  receiptHandle = message.ReceiptHandle;
-                } catch (e) {
-                  return callback(e);
-                }
-
-                processMessage(parsedMessage, (err) => {
-                  // We're just going to black-hole errors here
-                  if (ERR(err, (err) => logger.error('Error processing message', err)));
-
-                  const deleteParams = {
-                    QueueUrl: QUEUE_URL,
-                    ReceiptHandle: receiptHandle,
-                  };
-                  sqs.deleteMessage(deleteParams, (err) => {
-                    if (ERR(err, callback)) return;
-                    return callback(null);
-                  });
-                });
-              },
-              (err) => {
-                if (ERR(err, callback)) return;
-                return callback(null);
-              }
-            );
-          },
-        ],
-        (err) => {
-          if (ERR(err, (err) => logger.error('Error handling grader results', err)));
-          next();
+  const queueUrl = await loadQueueUrl(sqs);
+
+  // Start work in an IIFE so we can keep going asynchronously
+  // after we return to the caller.
+  (async () => {
```

Comment:
I ended up rewriting this file in `async`/`await`. The core thing we needed was the ability to actually wait for processing to finish; previously we would get results from SQS and then essentially fire-and-forget the actual processing. Now that we specifically need to wait for processing to stop, we need to propagate the notion of "done processing" back up. I could have done that wil callbacks, but it was only marginally more work to rewrite everything 🤷 

We'll have to make sure we test this carefully.